manual/io/index.md

```diff
@@ -56,784 +46,433 @@ root [1] TFile f1("my.root");
-The XML file is created in memory before it is stored on disk. As for normal ROOT files, XML files use the same I/O mechanism that the ROOT/Cling dictionary uses. Any class that has a dictionary can be stored in XML format. XML files do not support subdirectories or trees.
+The schema evolution C++ API consists of the follwing classes:
 
-To create an XML file, specify a filename with an `.xml` extension when calling [TFile::Open()](https://root.cern/doc/master/classTFile.html#ad8870806a04da2c2f4aa02bee4ec6833){:target="_blank"}. `TFile::Open()` recognizes that you are trying to open an XML file and returns a {% include ref class="TXMLFile" %} object. When a XML file is open in write mode, you can use [TObject::Write()](https://root.cern/doc/master/classTObject.html#a19782a4717dbfd4857ccd9ffa68aa06d){:target="_blank"} to write an object into the XML file.
+- [TSchemaRuleSet](https://root.cern/doc/master/classROOT_1_1Detail_1_1TSchemaRuleSet.html){:target="_blank"}: objects of this type manage the sets of rules and ensure their consistency. There can be no conflicting rules in the rule sets. The rule sets are owned by the {% include ref class="TClass" %} objects corresponding to the target classes defined in the rules and can be accessed using [`TClass::GetSchemaRules()`](https://root.cern/doc/master/classTClass.html#a8ffa89c0d9362cd77d957fed12ea9fca){:target="_blank"} and [`TClass::AdoptSchemaRules()`](https://root.cern/doc/master/classTClass.html#a484ed223d16ffbf0d35cee567e6f6225){:target="_blank"}.
+- [TSchemaRule](https://root.cern/doc/master/classROOT_1_1TSchemaRule.html){:target="_blank"}: it represent the rules and their fields have exactly the same meaning as the ones of rules specified in the dictionaries (see [above](#specifying-io-customization-rules-in-a-linkdef-file)).
 
-_**Example**_
+#### Schema evolution with custom streamers
+
+If you have written your own `Streamer` as described in [Custom streamers](#custom-streamers), you will have to manually add code for each version and manage the evolution of your class. When you add or remove data members, you must modify the `Streamer` by hand. ROOT assumes that you have increased the class version number in the `ClassDef` statement and introduced the relevant test in the read part of the Streamer. For example, if a new version of the `Event` class above includes a new member: `Int_t fNew` the `ClassDef` statement should be changed to `ClassDef(Event,2)` and the following lines should be added to the read part of the `Streamer`:
 
 {% highlight C++ %}
-// Example of a session saving a histogram to a XML file.
-   TFile *f = TFile::Open("Example.xml","recreate");
-   TH1F *h = new TH1F("h","test",1000,-2,2)
-   h->FillRandom("gaus");
-   h->Write();
-  delete f;
-
-// Example of a session saving a histogram to a XML file.
-   TFile *f = TFile::Open("Example.xml");
-   TH1F *h = (TH1F*)f->Get("h");
-   h->Draw();
-   canvas->Print("Example.xml");
+   if (R__v > 1) R__b >> fNew;
```

Comment:
I guess `R__b` is the `TBuffer` argument, but do we refer to `R__v` anywhere else?

Reply:
we don't say what `R__b` is, nor `R__v` (the version number). Rather than just _saying_ it I'd like to point to some reference or some code that shows what signature you can expect a streamer function to have. @pcanal help? :smile:

Reply:
For R__b, this is the argument to the `Streamer` function, usually declared by `ClassDef`:
```
virtual_keyword void Streamer(TBuffer &R__b) ...
```
R__v is a local variable usually assigned via:
```
   if (b.IsReading()) {
      UInt_t R__s, R__c;
      Version_t R__v = b.ReadVersion(&R__s, &R__c);
      ...... actual custom streamer (often delegating to the default streamer)
      b.CheckByteCount(R__s, R__c, TTree::IsA());
   }
```